### PR TITLE
Add validation for null array DataType

### PR DESCRIPTION
**Pull request by afazel**
##  opendistro-for-elasticsearch/security pull request intake form
_Please provide as much details as possible to get feedback/acceptance on your PR quickly_

 
 1. __Category:__ (_Enhancement, New feature, Bug fix, Test fix, Refactoring, Maintenance, Documentation_)
 Bug fix
 

 2. __Github Issue # or road-map entry, if available:__
#1136 

 3. __Description of changes:__
Add validation for all the API keys that have ARRAY DataType to make sure that there is no null value in the array.

 4. __Why these changes are required?__ 
By adding this validation, we prevent ES being completely unoperational when there are null in the arrays.

 5. __What is the old behavior before changes and new behavior after changes?__ (_Please add any example/logs/screen-shot if available_)
It was possible to bring ES completely unoperational by using null in the array DataTypes in several APIs.

 6. __Testing done:__ (_Please provide details of testing done: Unit testing, integration testing and manual testing_)
 Unit test added to cover array DataType like users with null in their list.
 
 
 7. __TO-DOs, if any:__ (_Please describe pending items and provide Github issues# for each of them_)

 8. __Is it backport from main branch?__ (_If yes, please add backport PR # and commits #_)

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/validation/AbstractConfigurationValidator.java [hunk 1]
```diff
@@ -167,6 +168,25 @@ public boolean validate() {
         mandatory.removeAll(requested);
         missingMandatoryKeys.addAll(mandatory);
 
+        //null element in the values of all the possible keys with DataType as ARRAY
+        final Map<String, Object> contentAsMap = XContentHelper.convertToMap(this.content, false, XContentType.JSON).v2();
```
Comment (vrozov) [inline src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/validation/AbstractConfigurationValidator.java:172]: There is a call to `DefaultObjectMapper.readTree()` on line 133. Can't it be reused here?

File: src/test/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/api/RolesMappingApiTest.java [hunk 2]
```diff
@@ -392,4 +392,16 @@ public void testRolesMappingApiForNonSuperAdmin() throws Exception {
...
+	@Test
+	public void checkNullElementsInArray() throws Exception{
+		setup();
+		rh.keystore = "restapi/kirk-keystore.jks";
+		rh.sendAdminCertificate = true;
+
+		HttpResponse response = rh.executePutRequest("/_opendistro/_security/api/rolesmapping/opendistro_security_role_starfleet_captains",
+													 FileHelper.loadFile("restapi/rolesmapping_null_array_element.json"), new Header[0]);
+		System.out.println(response.getBody());
+		Assert.assertTrue(response.getBody().contains("Invalid array element"));
```
Comment (sujithvm) [inline src/test/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/api/RolesMappingApiTest.java:405]: nit: Can we assert for `SC_BAD_REQUEST` here ? 
Also would be good to add tests for other APIs accepting array type

File: src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/validation/AbstractConfigurationValidator.java (hunk 1 above)
Comment (afazel) [reply src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/validation/AbstractConfigurationValidator.java]: Thanks Vlad, I used this in my new commit

File: src/test/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/api/RolesMappingApiTest.java (hunk 2 above)
Comment (afazel) [reply src/test/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/api/RolesMappingApiTest.java]: Sure, I changed to `SC_BAD_REQUEST` and also added unit tests for other APIs.

**Comment by afazel** [comment]
The failure in the unit tests was related to the bug in my code, I fixed the bug in my last commit and now all the tests passed.

File: src/test/resources/restapi/actiongroup_null_array_element.json [hunk 3]
```diff
@@ -0,0 +1,3 @@
+{
+ "allowed_actions": [null]
```
Comment (vrozov) [inline src/test/resources/restapi/actiongroup_null_array_element.json:2]: add use case for more than one element in the array, one of which is null.

File: src/test/resources/restapi/rolesmapping_null_array_element.json [hunk 4]
```diff
@@ -0,0 +1,11 @@
+{
+  "users": [
+    null
+  ],
+  "backend_roles": [
```
Comment (vrozov) [inline src/test/resources/restapi/rolesmapping_null_array_element.json:5]: add test cases for `backend_roles` and `hosts`

Comment (afazel) [reply src/test/resources/restapi/rolesmapping_null_array_element.json]: Sure, done

File: src/test/resources/restapi/actiongroup_null_array_element.json [hunk 5]
```diff
@@ -0,0 +1,3 @@
...
+}
```
[the remark this answers is not in the document]
Comment (afazel) [reply src/test/resources/restapi/actiongroup_null_array_element.json]: Done

File: src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/validation/AbstractConfigurationValidator.java [hunk 6]
```diff
@@ -281,7 +312,8 @@ private void addErrorMessage(final XContentBuilder builder, final String message
     public static enum ErrorType {
         NONE("ok"), INVALID_CONFIGURATION("Invalid configuration"), INVALID_PASSWORD("Invalid password"), WRONG_DATATYPE("Wrong datatype"),
         BODY_NOT_PARSEABLE("Could not parse content of request."), PAYLOAD_NOT_ALLOWED("Request body not allowed for this action."),
-        PAYLOAD_MANDATORY("Request body required for this action."), SECURITY_NOT_INITIALIZED("Security index not initialized");
+        PAYLOAD_MANDATORY("Request body required for this action."), SECURITY_NOT_INITIALIZED("Security index not initialized"),
+        INVALID_ARRAY_ELEMENT("Invalid array element");
```
[the remark this answers is not in the document]
Comment (afazel) [reply src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/validation/AbstractConfigurationValidator.java]: Yeah, fixed it.

File: src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/validation/AbstractConfigurationValidator.java [hunk 7]
```diff
@@ -188,6 +190,31 @@ public boolean validate() {
             return false;
         }
 
+        //null element in the values of all the possible keys with DataType as ARRAY
+        try {
+            HashMap<String,Object> contentAsMap = DefaultObjectMapper.readValue(content.utf8ToString(), HashMap.class);
+            for (String allowedKey : allowedKeys.keySet()) {
```
[the remark this answers is not in the document]
Comment (afazel) [reply src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/validation/AbstractConfigurationValidator.java]: Done

File: src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/validation/AbstractConfigurationValidator.java [hunk 8]
```diff
@@ -188,6 +190,31 @@ public boolean validate() {
             return false;
         }
 
+        //null element in the values of all the possible keys with DataType as ARRAY
+        try {
+            HashMap<String,Object> contentAsMap = DefaultObjectMapper.readValue(content.utf8ToString(), HashMap.class);
```
[the remark this answers is not in the document]
Comment (afazel) [reply src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/validation/AbstractConfigurationValidator.java]: Done

File: src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/validation/AbstractConfigurationValidator.java [hunk 9]
```diff
@@ -281,7 +305,8 @@ private void addErrorMessage(final XContentBuilder builder, final String message
     public static enum ErrorType {
         NONE("ok"), INVALID_CONFIGURATION("Invalid configuration"), INVALID_PASSWORD("Invalid password"), WRONG_DATATYPE("Wrong datatype"),
         BODY_NOT_PARSEABLE("Could not parse content of request."), PAYLOAD_NOT_ALLOWED("Request body not allowed for this action."),
-        PAYLOAD_MANDATORY("Request body required for this action."), SECURITY_NOT_INITIALIZED("Security index not initialized");
+        PAYLOAD_MANDATORY("Request body required for this action."), SECURITY_NOT_INITIALIZED("Security index not initialized"),
+        NULL_ARRAY_ELEMENT("Null array element");
```
Comment (vrozov) [inline src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/validation/AbstractConfigurationValidator.java:309]: "\`null\` is not allowed as json array element"

File: src/test/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/api/ActionGroupsApiTest.java [hunk 10]
```diff
@@ -359,4 +359,15 @@ public void testActionGroupsApiForNonSuperAdmin() throws Exception {
...
     }
 
+    @Test
+    public void checkNullElementsInArray() throws Exception{
+        setup();
+        rh.keystore = "restapi/kirk-keystore.jks";
+        rh.sendAdminCertificate = true;
+
+        HttpResponse response = rh.executePutRequest("/_opendistro/_security/api/actiongroups/CRUD_UT",
+                                                     FileHelper.loadFile("restapi/actiongroup_null_array_element.json"), new Header[0]);
```
[the remark this answers is not in the document]
Comment (vrozov) [reply src/test/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/api/ActionGroupsApiTest.java:369]: Using file is more justified for extra long strings with lots of double quotes escaping. In this case it is a short string and using file adds extra level of indirection and reduce code readability.

[the remark this answers is not in the document]
Comment (afazel) [reply src/test/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/api/ActionGroupsApiTest.java]: Got it, fixed

File: src/test/resources/restapi/actiongroup_null_array_element.json (hunk 5 above)
[the remark this answers is not in the document]
Comment (vrozov) [reply src/test/resources/restapi/actiongroup_null_array_element.json]: Same for other newly added files

File: src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/validation/AbstractConfigurationValidator.java [hunk 11]
```diff
@@ -188,6 +189,24 @@ public boolean validate() {
             return false;
         }
 
+        //null element in the values of all the possible keys with DataType as ARRAY
+        JsonNode contentAsNode = getContentAsNode();
```
Comment (vrozov) [inline src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/validation/AbstractConfigurationValidator.java:193]: It is not necessary to introduce local variable, use member variable directly.

File: src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/validation/AbstractConfigurationValidator.java [hunk 12]
```diff
@@ -188,6 +189,24 @@ public boolean validate() {
...
+        for (Entry<String, DataType> allowedKey : allowedKeys.entrySet()) {
+            DataType dataType = allowedKey.getValue();
+            if (dataType == DataType.ARRAY && contentAsNode.findValue(allowedKey.getKey()) != null) {
```
Comment (vrozov) [inline src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/validation/AbstractConfigurationValidator.java:196]: Avoid double calls to `contentAsNode.findValue()`, also I don't think that `findValue()` is the right method. Use `get()`.

File: src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/validation/AbstractConfigurationValidator.java [hunk 13]
```diff
@@ -188,6 +189,24 @@ public boolean validate() {
...
+                JsonNode value = contentAsNode.findValue(allowedKey.getKey());
```
Comment (vrozov) [inline src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/validation/AbstractConfigurationValidator.java:197]: Cast to `JsonArray` as data type is known to be `DataType.ARRAY`.

File: src/test/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/api/RolesApiTest.java [hunk 14]
```diff
@@ -469,8 +469,9 @@ public void testRolesApiForNonSuperAdmin() throws Exception {
         Assert.assertEquals(HttpStatus.SC_FORBIDDEN, response.getStatusCode());
 
         // Put read only roles
-        response = rh.executePutRequest("/_opendistro/_security/api/roles/opendistro_security_transport_client", "[{ \"op\": \"replace\", \"path\": \"/description\", \"value\": \"foo\" }]", new Header[0]);
-        Assert.assertEquals(HttpStatus.SC_FORBIDDEN, response.getStatusCode());
+        response = rh.executePutRequest("/_opendistro/_security/api/roles/opendistro_security_transport_client",
+                                        FileHelper.loadFile("restapi/roles_captains.json"), new Header[0]);
+        Assert.assertEquals(org.apache.http.HttpStatus.SC_FORBIDDEN, response.getStatusCode());
```
Comment (vrozov) [inline src/test/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/api/RolesApiTest.java:474]: Is this change necessary? Please double check other places for similar changes.

File: src/test/resources/restapi/roles_null_array_element.json [hunk 15]
```diff
@@ -0,0 +1,9 @@
+{
+  "cluster_permissions" : [null],
+  "index_permissions" : [ {
+    "index_patterns" : [ "*" ],
```
Comment (vrozov) [inline src/test/resources/restapi/roles_null_array_element.json:4]: Add use case for `null` in "index_patterns", "masked_fields" and "allowed_actions" and any of those being `null` by itself.

File: src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/validation/AbstractConfigurationValidator.java [hunk 16]
```diff
@@ -188,6 +190,24 @@ public boolean validate() {
             return false;
         }
 
+        //null element in the values of all the possible keys with DataType as ARRAY
+        HashMap<String,Object> contentAsMap = DefaultObjectMapper.objectMapper.convertValue(getContentAsNode(), HashMap.class);
```
[the remark this answers is not in the document]
Comment (afazel) [reply src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/validation/AbstractConfigurationValidator.java]: fixed

File: src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/validation/AbstractConfigurationValidator.java (hunk 11 above)
Comment (afazel) [reply src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/validation/AbstractConfigurationValidator.java]: fixed

File: src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/validation/AbstractConfigurationValidator.java (hunk 12 above)
Comment (afazel) [reply src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/validation/AbstractConfigurationValidator.java]: fixed

File: src/test/resources/restapi/roles_null_array_element.json (hunk 15 above)
Comment (afazel) [reply src/test/resources/restapi/roles_null_array_element.json]: Done

File: src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/validation/AbstractConfigurationValidator.java [hunk 17]
```diff
@@ -188,6 +191,42 @@ public boolean validate() {
             return false;
         }
 
+        //null element in the values of all the possible keys with DataType as ARRAY
+        for (Entry<String, DataType> allowedKey : allowedKeys.entrySet()) {
+            DataType dataType = allowedKey.getValue();
```
Comment (vrozov) [inline src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/validation/AbstractConfigurationValidator.java:196]:
```suggestion
            if (allowedKey.getValue() == DataType.ARRAY) {
              JsonArray values = (JsonArray)contentAsNode.get(allowedKey.getKey());
```

File: src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/validation/AbstractConfigurationValidator.java [hunk 18]
```diff
@@ -188,6 +191,42 @@ public boolean validate() {
...
+            if (dataType == DataType.ARRAY && value != null) {
+                try {
+                    List contentArray = DefaultObjectMapper.objectMapper.convertValue(value, java.util.List.class);
+                    if (contentArray.contains(null)) {
+                        this.errorType = ErrorType.NULL_ARRAY_ELEMENT;
+                        return false;
+                    }
+                    else {
+                        for (int i = 0; i < contentArray.size(); i++) {
+                            if (contentArray.get(i) instanceof HashMap) {
```
Comment (vrozov) [inline src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/validation/AbstractConfigurationValidator.java:207]: What if it is more deeply nested structure? Can you work directly with `JsonNode` and `JsonArray`?

File: src/test/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/api/RolesApiTest.java [hunk 19]
```diff
@@ -489,8 +490,9 @@ public void testRolesApiForNonSuperAdmin() throws Exception {
         Assert.assertEquals(HttpStatus.SC_NOT_FOUND, response.getStatusCode());
 
         // put hidden role
-        response = rh.executePutRequest("/_opendistro/_security/api/roles/opendistro_security_internal", "[{ \"op\": \"replace\", \"path\": \"/description\", \"value\": \"foo\" }]", new Header[0]);
-        Assert.assertEquals(HttpStatus.SC_NOT_FOUND, response.getStatusCode());
+        response = rh.executePutRequest("/_opendistro/_security/api/roles/opendistro_security_internal",
+                                        FileHelper.loadFile("restapi/roles_captains.json"), new Header[0]);
+        Assert.assertEquals(org.apache.http.HttpStatus.SC_NOT_FOUND, response.getStatusCode());
```
Comment (vrozov) [inline src/test/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/api/RolesApiTest.java:495]: Please fix

File: src/main/java/org/opensearch/security/dlic/rest/validation/AbstractConfigurationValidator.java [hunk 20]
```diff
@@ -297,4 +316,25 @@ public String getMessage() {
     protected final boolean hasParams() {
         return param != null && param.length > 0;
     }
+
+    private boolean hasNullElement(JsonNode currentNode) {
+        if (currentNode.isArray()) {
+            List contentArray = DefaultObjectMapper.objectMapper.convertValue(currentNode, java.util.List.class);
```
Comment (vrozov) [inline src/main/java/org/opensearch/security/dlic/rest/validation/AbstractConfigurationValidator.java:322]: Is this conversion necessary? If it is an array, can you cast it to `ArrayNode`?

File: src/main/java/org/opensearch/security/dlic/rest/validation/AbstractConfigurationValidator.java [hunk 21]
```diff
@@ -188,6 +191,17 @@ public boolean validate() {
             return false;
         }
 
+        //null element in the values of all the possible keys with DataType as ARRAY
+        for (Entry<String, DataType> allowedKey : allowedKeys.entrySet()) {
+            DataType dataType = allowedKey.getValue();
+            JsonNode value = contentAsNode.get(allowedKey.getKey());
+            if (dataType == DataType.ARRAY && value != null) {
```
Comment (vrozov) [inline src/main/java/org/opensearch/security/dlic/rest/validation/AbstractConfigurationValidator.java:198]: Remove check for `dataType`. It may be a json node that has json array as an element. With nested check for `null` elements in json array it is not necessary to rely on `dataType`.

File: src/main/java/org/opensearch/security/dlic/rest/validation/AbstractConfigurationValidator.java [hunk 22]
```diff
@@ -297,4 +316,25 @@ public String getMessage() {
     protected final boolean hasParams() {
         return param != null && param.length > 0;
     }
+
+    private boolean hasNullElement(JsonNode currentNode) {
```
Comment (vrozov) [inline src/main/java/org/opensearch/security/dlic/rest/validation/AbstractConfigurationValidator.java:320]: nit: rename `currentNode` to `node`.

File: src/main/java/org/opensearch/security/dlic/rest/validation/AbstractConfigurationValidator.java [hunk 23]
```diff
@@ -297,4 +316,25 @@ public String getMessage() {
...
+        if (currentNode.isArray()) {
+            List contentArray = DefaultObjectMapper.objectMapper.convertValue(currentNode, java.util.List.class);
+            if(contentArray.contains(null)){
+                return true;
+            }
+            ArrayNode arrayNode = (ArrayNode) currentNode;
+            Iterator<JsonNode> node = arrayNode.elements();
+            while (node.hasNext()) {
+                if( hasNullElement(node.next())) return true;
```
Comment (vrozov) [inline src/main/java/org/opensearch/security/dlic/rest/validation/AbstractConfigurationValidator.java:329]: nit: `if (hasNullElement`. While we don't enforce any particular code style, it will be better to make it consistent with the rest of the file. It is very uncommon to have space after `(` in the `if` condition.

File: src/main/java/org/opensearch/security/dlic/rest/validation/AbstractConfigurationValidator.java [hunk 24]
```diff
@@ -297,4 +316,25 @@ public String getMessage() {
...
+        }
+        else if (currentNode.isObject()) {
+            Iterator<Entry<String, JsonNode>> entry_it = currentNode.fields();
+            while (entry_it.hasNext()) {
+                if (hasNullElement(entry_it.next().getValue())) return true;
+            }
+        }
+        return false;
+    }
 }
```
Comment (vrozov) [inline src/main/java/org/opensearch/security/dlic/rest/validation/AbstractConfigurationValidator.java:340]: nit: add new line

Comment (afazel) [reply src/main/java/org/opensearch/security/dlic/rest/validation/AbstractConfigurationValidator.java]: Done

File: src/main/java/org/opensearch/security/dlic/rest/validation/AbstractConfigurationValidator.java (hunk 23 above)
Comment (afazel) [reply src/main/java/org/opensearch/security/dlic/rest/validation/AbstractConfigurationValidator.java]: Done

File: src/main/java/org/opensearch/security/dlic/rest/validation/AbstractConfigurationValidator.java (hunk 22 above)
Comment (afazel) [reply src/main/java/org/opensearch/security/dlic/rest/validation/AbstractConfigurationValidator.java]: Done

File: src/main/java/org/opensearch/security/dlic/rest/validation/AbstractConfigurationValidator.java (hunk 21 above)
Comment (afazel) [reply src/main/java/org/opensearch/security/dlic/rest/validation/AbstractConfigurationValidator.java]: Done

File: src/main/java/org/opensearch/security/dlic/rest/validation/AbstractConfigurationValidator.java (hunk 20 above)
Comment (afazel) [reply src/main/java/org/opensearch/security/dlic/rest/validation/AbstractConfigurationValidator.java]: Used `ArrayNode` in the latest commit

File: src/main/java/org/opensearch/security/dlic/rest/validation/AbstractConfigurationValidator.java (hunk 21 above)
Comment (afazel) [reply src/main/java/org/opensearch/security/dlic/rest/validation/AbstractConfigurationValidator.java]: Done

File: src/test/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/api/RolesApiTest.java (hunk 19 above)
Comment (afazel) [reply src/test/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/api/RolesApiTest.java]: Done

File: src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/validation/AbstractConfigurationValidator.java (hunk 18 above)
Comment (afazel) [reply src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/validation/AbstractConfigurationValidator.java]: Done

File: src/main/java/org/opensearch/security/dlic/rest/validation/AbstractConfigurationValidator.java [hunk 25]
```diff
@@ -297,4 +315,26 @@ public String getMessage() {
...
+
+    private boolean hasNullElement(JsonNode node) {
+        if (node.isArray()) {
+            for (JsonNode jsonNode: node){
+                if(jsonNode.isNull()){
+                    return true;
+                }
+            }
+            ArrayNode arrayNode = (ArrayNode) node;
+            Iterator<JsonNode> nodeIter = arrayNode.elements();
```
Comment (vrozov) [inline src/main/java/org/opensearch/security/dlic/rest/validation/AbstractConfigurationValidator.java:327]: It is confusing in a single method to use `node.iterator()`, `node.elements()` and `node.fields()`. Please unify to use `node.iterator()`.

Comment (afazel) [reply src/main/java/org/opensearch/security/dlic/rest/validation/AbstractConfigurationValidator.java:327]: I am not sure how should I do it, as it depends whether the `JsonNode` is `array` or `object`. I can not see any confusion as the if statements are clear enough to see the type of the JsonNode and how we are handling it.

Comment (vrozov) [reply src/main/java/org/opensearch/security/dlic/rest/validation/AbstractConfigurationValidator.java:327]: Won't `node.iterator()` (`for (JsonNode jsonNode: node) {}`) work both for `array` and `object`? Why is it necessary to use `node.fields()` for `object`?

Comment (afazel) [reply src/main/java/org/opensearch/security/dlic/rest/validation/AbstractConfigurationValidator.java]: Done

Comment (afazel) [reply src/main/java/org/opensearch/security/dlic/rest/validation/AbstractConfigurationValidator.java:326]: I tried to use only `node.iterator` but it cause failure in some other tests like `AuditApiActionTest`, debugging it showed that it can not handle all the objects.

Comment (vrozov) [reply src/main/java/org/opensearch/security/dlic/rest/validation/AbstractConfigurationValidator.java]: Please provide details why `node.iterator()` can not handle all objects.

Comment (afazel) [reply src/main/java/org/opensearch/security/dlic/rest/validation/AbstractConfigurationValidator.java]: We just want to make sure that an array element does not have a null, right? This code:
```
for (JsonNode jsonNode : node) {
      if(jsonNode.isNull()){
         return true;
       }
       else {
            Iterator<JsonNode> nodeIter = jsonNode.iterator();
             while (nodeIter.hasNext()) {
                        if (hasNullElement(nodeIter.next()))
                            return true;
              }
       }
}
```
return `true` for inputs like this:
```{"enabled":true,"write_log_diffs":false,"read_watched_fields":null,"read_ignore_users":[],"write_watched_indices":[],"write_ignore_users":["kibanaserver"],"read_metadata_only":true,"write_metadata_only":false,"external_config":true,"internal_config":true}```

Because the value of `read_watched_fields` is `null`, however it is ok, since `read_watched_fields` is not an array.
What is wrong with the implementation in the last commit? I made it more clear and it is really straight forward.

Comment (afazel) [reply src/main/java/org/opensearch/security/dlic/rest/validation/AbstractConfigurationValidator.java]: OK, I modified `hasNullElement()` method to only use node iterator in my latest commit.. Hopefully it is good now.

File: src/main/java/org/opensearch/security/dlic/rest/validation/AbstractConfigurationValidator.java [hunk 26]
```diff
@@ -22,6 +22,7 @@
 import java.util.Map;
 import java.util.Map.Entry;
 import java.util.Set;
+import java.util.Iterator;
```
Comment (vrozov) [inline src/main/java/org/opensearch/security/dlic/rest/validation/AbstractConfigurationValidator.java:25]: Check for unused imports

File: src/main/java/org/opensearch/security/dlic/rest/validation/AbstractConfigurationValidator.java [hunk 27]
```diff
@@ -297,4 +314,16 @@ public String getMessage() {
     protected final boolean hasParams() {
         return param != null && param.length > 0;
     }
-}
\ No newline at end of file
+
+    private boolean hasNullElement(JsonNode node) {
```
Comment (vrozov) [inline src/main/java/org/opensearch/security/dlic/rest/validation/AbstractConfigurationValidator.java:318]: rename to `hasNullArrayElement`

File: src/main/java/org/opensearch/security/dlic/rest/validation/AbstractConfigurationValidator.java [hunk 28]
```diff
@@ -297,4 +314,16 @@ public String getMessage() {
...
+        for (JsonNode jsonNode: node) {
```
Comment (vrozov) [inline src/main/java/org/opensearch/security/dlic/rest/validation/AbstractConfigurationValidator.java:319]: nit: `jsonNode` -> `element`

File: src/main/java/org/opensearch/security/dlic/rest/validation/AbstractConfigurationValidator.java [hunk 29]
```diff
@@ -297,4 +314,16 @@ public String getMessage() {
...
+            if(jsonNode.isNull() && node.isArray()) {
```
Comment (vrozov) [inline src/main/java/org/opensearch/security/dlic/rest/validation/AbstractConfigurationValidator.java:320]: If `jsonNode` is `null` and `node` is not an array, it is not necessary to call `hasNullElement`. We need to call `hasNullElement()` only if `jsonNode` is object or an array:
```
if (jsonNode.isNull()) {
  if (node.isArray()) {
    return true;
  }
} else if (jsonNode.isContainerNode()) {
  if (hasNullElement(jsonNode)) {
    return true;
  }
}
```

File: src/test/java/org/opensearch/security/dlic/rest/api/RolesApiTest.java [hunk 30]
```diff
@@ -489,7 +489,8 @@ public void testRolesApiForNonSuperAdmin() throws Exception {
         Assert.assertEquals(HttpStatus.SC_FORBIDDEN, response.getStatusCode());
 
         // Put read only roles
-        response = rh.executePutRequest(ENDPOINT + "/roles/opendistro_security_transport_client", "[{ \"op\": \"replace\", \"path\": \"/description\", \"value\": \"foo\" }]", new Header[0]);
+        response = rh.executePutRequest("/_opendistro/_security/api/roles/opendistro_security_transport_client",
```
Comment (vrozov) [inline src/test/java/org/opensearch/security/dlic/rest/api/RolesApiTest.java:492]: incorrect merge during rebase

File: src/test/java/org/opensearch/security/dlic/rest/api/RolesApiTest.java [hunk 31]
```diff
@@ -509,8 +510,9 @@ public void testRolesApiForNonSuperAdmin() throws Exception {
         Assert.assertEquals(HttpStatus.SC_NOT_FOUND, response.getStatusCode());
 
         // put hidden role
-        response = rh.executePutRequest(ENDPOINT + "/roles/opendistro_security_internal", "[{ \"op\": \"replace\", \"path\": \"/description\", \"value\": \"foo\" }]", new Header[0]);
-        Assert.assertEquals(HttpStatus.SC_NOT_FOUND, response.getStatusCode());
+        String body = FileHelper.loadFile("restapi/roles_captains.json");
+        response = rh.executePutRequest("/_opendistro/_security/api/roles/opendistro_security_internal", body, new Header[0]);
```
Comment (vrozov) [inline src/test/java/org/opensearch/security/dlic/rest/api/RolesApiTest.java:514]: incorrect merge during rebase

File: src/main/java/org/opensearch/security/dlic/rest/validation/AbstractConfigurationValidator.java (hunk 27 above)
Comment (afazel) [reply src/main/java/org/opensearch/security/dlic/rest/validation/AbstractConfigurationValidator.java]: Done

File: src/main/java/org/opensearch/security/dlic/rest/validation/AbstractConfigurationValidator.java (hunk 29 above)
Comment (afazel) [reply src/main/java/org/opensearch/security/dlic/rest/validation/AbstractConfigurationValidator.java]: fixed

File: src/test/java/org/opensearch/security/dlic/rest/api/RolesApiTest.java (hunk 30 above)
Comment (afazel) [reply src/test/java/org/opensearch/security/dlic/rest/api/RolesApiTest.java]: fixed

File: src/test/java/org/opensearch/security/dlic/rest/api/RolesApiTest.java (hunk 31 above)
Comment (afazel) [reply src/test/java/org/opensearch/security/dlic/rest/api/RolesApiTest.java]: fixed

File: src/main/java/org/opensearch/security/dlic/rest/validation/AbstractConfigurationValidator.java [hunk 32]
```diff
@@ -297,4 +312,18 @@ public String getMessage() {
     protected final boolean hasParams() {
         return param != null && param.length > 0;
     }
-}
\ No newline at end of file
+
+    private boolean hasNullArrayElement(JsonNode node) {
+        for (JsonNode element: node) {
+            if(element.isNull()) {
+                if (node.isArray())
```
Comment (vrozov) [inline src/main/java/org/opensearch/security/dlic/rest/validation/AbstractConfigurationValidator.java:319]: while code style is not explicitly enforced it will be better to adhere to what is mostly used by the rest of files. `if` must have `{}` block with `{` on the same line as `if` and `}` on the same line as `else if (`. Please see my code example for the reference.

Comment (afazel) [reply src/main/java/org/opensearch/security/dlic/rest/validation/AbstractConfigurationValidator.java]: Fixed

**Review by vrozov — APPROVED** [review]

**Review by hsiang9431-amzn — APPROVED** [review]
